pkg/game/game.go

```diff
@@ -0,0 +1,172 @@
+package game
+
+import (
+	"errors"
+	"io"
+
+	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game/elements"
+	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/logger"
+	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/stack"
+	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tiles"
+	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tilesets"
+)
+
+type Game struct {
+	board         elements.Board
+	deck          *stack.Stack[tiles.Tile]
+	players       []elements.Player
+	currentPlayer int
+	log           *logger.Logger
+}
+
+func NewGame(log *logger.Logger) (*Game, error) {
+	deck := stack.New(tilesets.GetStandardTiles())
+	return NewGameWithDeck(&deck, log)
+}
+
+func NewGameWithDeck(
+	deck *stack.Stack[tiles.Tile], log *logger.Logger,
+) (*Game, error) {
+	if log == nil {
+		nullLogger := logger.New(io.Discard)
+		log = &nullLogger
+	}
+	game := &Game{
+		board:         NewBoard(deck.GetTileSet()),
+		deck:          deck,
+		players:       []elements.Player{NewPlayer(0), NewPlayer(1)},
+		currentPlayer: 0,
+		log:           log,
+	}
+
+	// All tiles in base game can be placed on the first move but let's just check this
+	// in case this isn't true for tiles from all of the expansions.
+	err := game.ensureCurrentTileHasValidPlacement()
+	if err != nil {
+		return nil, err
+	}
+	if err := log.LogEvent(
+		logger.NewStartEntry(game.deck, len(game.players)),
+	); err != nil {
+		return nil, err
+	}
+
+	return game, nil
+}
+
+func (game *Game) GetCurrentTile() (tiles.Tile, error) {
+	return game.deck.Peek()
+}
+
+func (game *Game) CurrentPlayer() elements.Player {
+	return game.players[game.currentPlayer]
+}
+
+func (game *Game) PlayerCount() int {
+	return len(game.players)
+}
+
+func (game *Game) ensureCurrentTileHasValidPlacement() error {
+	for {
+		// Peek at the tile that will be returned by GetCurrentTile() next time
+		// to see, if it can actually placed anywhere.
+		nextTile, err := game.deck.Peek()
+		if err != nil {
+			if errors.Is(err, stack.ErrStackOutOfBounds) {
+				// out of bounds - not our concern at the end of a turn
+				return nil
+			}
+
+			// unexpected error?
```

Comment:
Maybe Panic?

Reply:
Will we be catching panics in game engine or could this crash the whole process during costly machine learning when we could potentially recover (by ignoring that one game during learning process and logging the isssue)?

Reply:
Since we do not know the error it may as well crash the whole process.